topostats/filters.py

```diff
@@ -290,7 +291,7 @@ def gaussian_filter(self, image: np.ndarray, **kwargs) -> np.array:
             **kwargs,
         )
 
-    def filter_image(self) -> None:
+    def filter_image(self, scars_config: dict) -> None:
```

Comment:
When setting out writing the `Scars` class, I considered this. I decided against it because I thought it would make the `filters.py` file even more massive, and far more difficult to develop / comprehend. 

I also decided against inheritance because this also would increase complexity. Also doing it this way gives us a little freedom to use the `Scars` class in a more modular way in the future if needs be - we might want to run scar removal independently of `filters`, which would be made easier with the modular way of implementing it - though perhaps I'm overdoing the future-proofing?

I agree that the method for passing the `scars` config seems a little odd on the surface, however the reason for including the `scars` input there, is that the `Scars` class needs to be instantiated during the filtering process, and passing multiple sets of `kwargs` is inadvisable in `init()` methods. 

What do you think? 😄

Reply:
@ns-rse what do you think about this?

Reply:
I think keeping `Scars` as its own class is eminently sensible.

One approach of using inheritance (and I haven't used it much at all) might be to have this pattern, passing the arguments to Scars appropriately using `super()`.

```
class Filters(Scars):
    def __init__(...):
        super().__init__(...)
```

That is `Filters` inherits `Scars` because you want filtering to have the functionality to remove scars.

I don't think this would necessarily require moving configuration sections around as you could simply combine the two dictionaries and pass them in when instantiating `Filters()`.

```
filter_config.update(scars_config)
```

Reply:
Thank you for that comprehensive analysis @ns-rse. Given that it *might* be possible that we want to have values derived from the filtering process affect parameters in scar removal, the current solution, while weird might be the simplest and least likely to cause problems for future development. 

@MaxGamill-Sheffield what are your thoughts about this?

Reply:
I agree that it's easier to understand and use but disagree about it being used outside of filters. Filters is essentially our pre-processing stage to end up with a suitable image we can then perform analysis on. Although, called filters, I believe that scar removal is included in this pre-processing pipeline.

With the modularisation of the filters class, I think that once the Filters object is initialised, code-competent users should be able to build their own pipelines using the functions within it (flatten then align rows vs align rows then flatten etc), in which I still think scars should belong. This might also be helped with an interactive GUI that we are looking into in collaboration with comp sci, where people could drag-and-drop modules (in this case our funcs) to build pipelines (possibly very far into the future however).

For those reasons, I like the idea of class inheritance but I can understand this is unfamiliar territory and may take a while to implement and we might want to avoid it, but this would help to remove the additional config input which I do think should be part of the filters config. Can talk about this on Tuesday.

Reply:
@ns-rse @MaxGamill-Sheffield should I refactor this so that the Scars class is a child of Filters?

Reply:
I think it would make sense as then we wouldn't have to pass it separate params and can include it in the filters workflow but I understand that it could slow down development, unless this isn't too difficult of a task and doesn't require much refactoring but I wouldn't know as I'm also not familiar with child classes.

If it does, you could bring it up with Alice in tomorrow's meeting?

Reply:
I'll give it a go!

Reply:
My current, poorly informed thinking, is as follows...

1. "filtering"  is the overall first pre-processing step (its actually many steps).
2. Its implemented in `Filters()` and currently has methods for `align_rows()` , `median_flatten()` etc.
3. We wish to add additional steps, i.e. scar removal.
4. That has been implemented in the `Scars()` class.
5. This additional functionality should be available _within_ the `Filters()` class.
6. There may be additional pre-processing steps that get implemented that should be available under `Filters()` (e.g. a class for `BlobRemoval()`).
7. `Filters()` should then inherit those, then when people wish to have control over the steps in which they process things they only need to instantiate `Filters()` and the functionality of `Scars()` / `BlobRemoval()` / `SomethingElseFunky()` are available in that instantiated object as they are all.

This could be done via inheritance but we're not dealing with representations of objects that have similar properties that inherit properties from a parent as in many examples of inheritance. 

To this end must apologise as I feel I may have wasted time and key-strokes by using the wrong term/design pattern as what I should have suggested is a design pattern known as composition as `Filters()` would be a composite of its existing methods, `Scars()`, `BlobRemoval()`, `SomethingElseFunky()` and anything else that is required.

An exposition of this is given in [Inheritance and Composition: A Python OOP Guide – Real Python](https://realpython.com/inheritance-composition-python/)

Some (crude) pseudo-code might be...

```
class Filters()
    def __init__(self, ...):
        ... <other options>
        self.scars = None
        self.blob_removal = None

    ... <all existing Filters methods>

    def filter_image(self) -> None:
        self.images["initial_median_flatten"] = self.median_flatten(self.images["pixels"], mask=None)
        ... <other filter steps required>
        if self.scars is not None:
            self.images["scars_removed"] = self.scars.remove_scars()
        if self.blob_removal is not None:
            self.images["blobs_removed"] = self.blob_removal.remove_blobs()
        ... <remaining filter steps after scar/blob removal> 

filters = Filters(**filter_config)
if scar_config["run"]:
    scar_config.pop("run")
    filters.scars = Scars(**scar_config)
if blob_config["run"]:
    blob_config.pop("run")
    filters.blob_removal = BlobRemoval(**blob_config)
# Make it so!
filters.filter_image()
```

Reply:
Change obsolete due to `scars.py` not being OOP now.